src/sentry/replays/lib/storage.py

```diff
@@ -0,0 +1,161 @@
+                    "segment_id": segment.segment_id,
+                },
+            )
+
+            # Delete the file from remote storage since it is un-retrievable.
+            file.delete()
+
+
+class StorageBlob(Blob):
+    """Storage service driver blob manager.
+
+    This driver does not have managed TTLs.  To enable TTLs you will need to enable it on your
+    bucket.  Keys are prefixed by their TTL.  Those TTLs are 30, 60, 90.  Measured in days.
+    """
+
+    def delete(self, segment: RecordingSegmentStorageMeta) -> None:
+        storage = get_storage(self._make_storage_options())
+        storage.delete(self.make_key(segment))
+
+    @metrics.wraps("replays.lib.storage.StorageBlob.get")
+    def get(self, segment: RecordingSegmentStorageMeta) -> bytes:
+        cache_key = make_filename(segment)
+
+        # Return eagerly from cache if the value exists.
+        cache_value = replay_cache.get(cache_key, raw=True)
+        if cache_value:
+            return cache_value
+
+        try:
+            storage = get_storage(self._make_storage_options())
+            blob = storage.open(self.make_key(segment))
+            result = blob.read()
+            blob.close()
+        except Exception:
+            # Return a default value if the storage does not exist.
+            return b"[]"
+        else:
+            # Set the returned value in cache.
+            replay_cache.set(cache_key, result, timeout=3600, raw=True)
+            return result
+
+    @metrics.wraps("replays.lib.storage.StorageBlob.set")
+    def set(self, segment: RecordingSegmentStorageMeta, value: bytes) -> None:
+        storage = get_storage(self._make_storage_options())
+        storage.save(self.make_key(segment), BytesIO(value))
+
+    def make_key(self, segment: RecordingSegmentStorageMeta) -> str:
+        return make_filename(segment)
+
+    def _make_storage_options(self) -> Optional[dict]:
+        backend = options.get("replay.storage.backend")
+        if backend:
+            return {"backend": backend, "options": options.get("replay.storage.options")}
+
+
+def make_filename(segment: RecordingSegmentStorageMeta) -> str:
+    """Return a deterministic segment filename.
+
+    Filename prefixes have special ordering requirements.  The first prefix "retention days" is
+    used for managing TTLs.  The project_id and replay_id prefixes are used for managing user
+    and GDPR deletions.
+    """
+    return "{}/{}/{}/{}".format(
+        segment.retention_days,
+        segment.project_id,
+        segment.replay_id,
+        segment.segment_id,
+    )
+
+
+def make_storage_driver(organization_id: int) -> Union[FilestoreBlob, StorageBlob]:
+    """Return a storage driver instance."""
+    if organization_id % 100 < options.get("replay.storage.direct-storage-sample-rate"):
+        return StorageBlob()
+    elif organization_id in settings.SENTRY_REPLAYS_STORAGE_ALLOWLIST:
+        return StorageBlob()
+    else:
+        return FilestoreBlob()
```

Comment:
do `StorageBlob` and `FilestoreBlob` need to be classes that are initialized per request given there's no instance data?

Reply:
Initialized as singletons.